proc-macro: Add support for fallible functions (#1408)
mozilla/uniffi-rs: docs/manual/src/proc_macro/index.md

}
```

## The `uniffi::Error` derive

> The `Error` derive registers a type as an error that be the `E` in a `Result<T, E>` return type of
> a `#[uniffi::export]`ed function or method. Used on its own, it can be used on any enum that the
> `Enum` derive also accepts, and exposes any variant fields to the foreign code. The generated
> foreign function will have the result's `T` as its return type, and throw the error in case the Rust
> call returns `Err(e)`.

Reviewer: This reads a bit weird.

Maybe we can rephrase that.

```suggestion
The `Error` derive registers a type as an error and can be used on any enum that the `Enum` derive also accepts.
It exposes any variant fields to the foreign code. 
This type can then be used as the `E` in a `Result<T, E>` return type of an exported function or method.
The generated foreign function for an exported function with a `Result<T, E>` return type
will have the result's `T` as its return type and throw the error in case the Rust call returns `Err(e)`.
```

I tried to make it more concise and re-order the bits a bit, but feel free to adjust that again.

(fwiw I generally recommend [semantic linebreaks](https://sembr.org/), makes further editing and reviewing easier most of the time.)

```rust
#[derive(uniffi::Error)]
pub enum MyError {
    MissingInput,
    IndexOutOfBounds {
        index: u32,
        size: u32,
    }
    Generic {
        message: String,
    }
}

#[uniffi::export]
fn do_thing() -> Result<(), MyError> {
    // ...
}
```

> You can also use the helper attribute `#[uniffi(flat_error)]` to expose just the variants but none

Author: Oops 😅

> You can also use the helper attribute `#[uniffi(flat_error)]` to expose just the variants but none
> of the fields, in this case the error will be serialized using Rust's `ToString` trait and will be
> accessible as the only field on each of the variants. In this case, your variants can have unnamed
> fields, and the types of the fields don't need to implement any special traits.

Reviewer: ```suggestion
You can also use the helper attribute `#[uniffi(flat_error)]` to expose just the variants but none
of the fields.
In this case the error will be serialized using Rust's `ToString` trait 
and will be accessible as the only field on each of the variants. 
For flat errors your variants can have unnamed fields, and the types of the fields don't need to implement any special traits.
```
